Mage.Sets/src/mage/cards/z/ZirdaTheDawnwaker.java

```diff
@@ -79,7 +78,11 @@ public boolean isLegal(Set<Card> deck, int minimumDeckSize) {
                 .allMatch(card -> card
                         .getAbilities()
                         .stream()
-                        .anyMatch(ActivatedAbility.class::isInstance)
+                        .anyMatch(ability -> // This may be imperfect? Not sure everything is excluded correctly.
+                                ability instanceof ActivatedAbility &&
+                                        !(ability instanceof SpellAbility) &&
+                                        !(ability instanceof PlayLandAbility) &&
+                                        !(ability instanceof SpecialAction))
```

Comment:
What you want to do here?

Reply:
Oh right SpecialAction don't have the inheritence problem of the other two that are extending ActivatedAbilityImpl. No idea why I added that.

Reply:
Oh no they have the same issue:
![image](https://github.com/magefree/mage/assets/34709007/fd2d4931-afad-4a1a-8881-9d39e234022a)

Reply:
Basically, Zirda need to find the activated abilities according to the rules in card, but the 3 Class I excluded there are not, and the previous filtering was finding that any permanent card was legal (because it had either SpellAbility or PlayLandAbility), and Zirda could be made Companion all the time. Which was incorrect.

Reply:
Just realized, should check `AbilityType` for `Activated` or `Mana` instead.

Reply:
That's?

![shot_240419_023428](https://github.com/magefree/mage/assets/8344157/b53f2d96-d121-4dee-abbc-8b3d3bc0015f)

Reply:
yep exactly (MANA too)